File: Leetcode_solutions/remove-linked-list-elements.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} val
 * @return {ListNode}
 */
var getNext = function(head, val) {
  let next = head;
  while (next && next.val === val) {
    next = next.next;
  }
  return next;
};

var removeElements = function(head, val) {
  head = getNext(head, val);
  let start = head,
    prev = head,
    next;
  while (prev) {
    next = prev.next;
    next = getNext(next, val);
    prev.next = next;
    prev = next;
  }
  return start;
};
console.log(removeElements([1, 2, 6, 3, 4, 5, 6], 6));

File: Leetcode_solutions/Remove Linked List Elements.py
class Solution:
    def removeElements(self, head: ListNode, val: int) -> ListNode:
        sentinel = ListNode(0)
        sentinel.next = head

        prev, curr = sentinel, head
        while curr:
            if curr.val == val:
                prev.next = curr.next
            else:
                prev = curr
            curr = curr.next

        return sentinel.nexta


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def removeElements(self, head: ListNode, val: int) -> ListNode:
        dummy = ListNode(0)
        dummy.next = head
        first = dummy
        second = head
        while second:
            if second.val == val:
                first.next = second.next
            else:
                first = second
            second = second.next
        return dummy.next
